raiden-ts/src/services/epics.ts

```diff
@@ -429,15 +442,133 @@ export const monitorUdcBalanceEpic = (
   {}: Observable<RaidenAction>,
   {}: Observable<RaidenState>,
   { address, latest$, userDepositContract }: RaidenEpicDeps,
-): Observable<udcDeposited> =>
+): Observable<udcDeposit.success> =>
   latest$.pipe(
     pluckDistinct('state', 'blockNumber'),
     // it's seems ugly to call on each block, but UserDepositContract doesn't expose deposits as
     // events, and ethers actually do that to monitor token balances, so it's equivalent
     exhaustMap(() => userDepositContract.functions.effectiveBalance(address) as Promise<UInt<32>>),
-    distinctUntilChanged((x, y) => y.eq(x)),
-    map(udcDeposited),
+    withLatestFrom(latest$),
+    filter(([balance, { udcBalance }]) => !udcBalance.eq(balance)),
+    map(([balance]) => udcDeposit.success(undefined, { totalDeposit: balance })),
+  );
+
+function makeUdcDeposit$(
+  [tokenContract, userDepositContract]: [HumanStandardToken, UserDeposit],
+  [sender, address]: [Address, Address],
+  [deposit, totalDeposit]: [UInt<32>, UInt<32>],
+  { pollingInterval, minimumAllowance }: RaidenConfig,
+  { log }: Pick<RaidenEpicDeps, 'log'>,
+) {
+  return defer(() =>
+    Promise.all([
+      tokenContract.functions.balanceOf(sender),
+      tokenContract.functions.allowance(sender, userDepositContract.address),
+      userDepositContract.functions.effectiveBalance(address),
+    ]),
+  ).pipe(
+    mergeMap(([balance, allowance, deposited]) => {
+      assert(deposited.add(deposit).eq(totalDeposit), [
+        ErrorCodes.UDC_DEPOSIT_OUTDATED,
+        { requested: totalDeposit.toString(), current: deposited.toString() },
+      ]);
+      assert(balance.gte(deposit), [
+        ErrorCodes.RDN_INSUFFICIENT_BALANCE,
+        { current: balance.toString(), required: deposit.toString() },
+      ]);
+
+      if (allowance.gte(deposit)) return of(true); // if allowance already enough
```

Comment:
This caused me issues. It is always hard to follow these return statements. There is this fundamental thing in my head that return statement end a function and go back on the heap. Instead I had to remember myself that I'm in a mergeMap, jump to its end to see what happens (because the functions states to deposit, which didn't happen yet) and then come back to checkout the else case. Could we please try to export in-themselves closed maps so we have an easy to follow flow? So I could read: first we do the approve (if necessary) then the deposit. Then this statement in a dedicated function just for the allowance/approval would not cause these issues I explained.

Reply:
Sorry if it may be hard for you to read these, but early `return` and early `throw` is essential in our workflow to _fail_ or _succeed_ early, in the sense of letting the bulk of operations in the end and with -1 indentation level as would be required. This is a pattern used thorough the codebase already, but we may discuss if it isn't reasonable and decide on a refactor.
I'll extract the `mergeMap`'s callback on these examples, but refactor patterns like these may be worth their own issue and proper discussion.

Reply:
Done

Reply:
Thanks. I know these comments are hard. :see_no_evil: